### DISCUSSION TITLE
node.js는 왜 싱글스레드인가요?

### DISCUSSION THREAD

**Discussion by michoball** [wanted-9th-3team/frontend-study#99] [프론트 기술 질문 공간]
자바스크립트는 싱글스레드 언어입니다. 

그럼 node.js 는 왜 싱글스레드로 작동하나요?  
자바스크립트가 싱글스레드로 만들어진 이유도 같이 살펴보면 좋을 것 같습니다.

**Comment by michoball** [comment]
우선 자바스크립트는 싱글 쓰레드 언어입니다. 그리고 그 이유는 바로 "쉽다" 라는 점 때문이라고 합니다. 
웹 페이지를 이용하면서 여러 작업들을 동시적으로 진행하고자 하는 경우가 있습니다. 그런 경우 자바스크립트가 멀티 쓰레드라면 동시 진행하는 작업들의 다양한 상황에 대처하기위해 많은 작업들을 해줘야할 것입니다.  하지만 싱글쓰레드언엉이기 때문에 이부분을 생각하지 않고 이벤트 루프 환경에 대해서만 인지를 잘하고 있으면 되서 자바스크립트는 싱글 쓰레드 언어입니다. 

node.js는 따지고보면 멀티 쓰레드라고 합니다. 하지만 메인 쓰레드가 돌아가는 환경이 js의 이벤트 루프이기 때문에 싱글쓰레드라고 한다고 합니다. 
자바스크립트가 비동기작업을 수행할때  Node api를 통해 libuv 라이브러리를 사용하여 처리를 한다고 합니다.  이때 이 libuv 라이브러리가 비동기처리를 위해 제공하는 것이 이벤트 루프라고 하네요. 

추가적으로 chat gpt 에게 물어본 결과도 첨부합니다. 

이벤트 루프 (Event Loop):
Node.js는 이벤트 루프를 사용하여 비동기 작업을 처리합니다. 이벤트 루프는 싱글 쓰레드에서 실행되며, 이벤트 기반으로 작동합니다. 이러한 설계는 많은 동시 요청을 처리하면서도 쓰레드 관리 및 동기화 문제를 줄일 수 있게 해줍니다.

비동기 I/O 작업:
Node.js는 주로 비동기 I/O 작업에 사용됩니다. 이것은 파일 시스템 액세스, 네트워크 요청 및 데이터베이스 쿼리와 같은 작업을 의미합니다. 비동기 작업은 블로킹되지 않으며, 이벤트 루프를 통해 비동기 작업이 완료되면 해당 작업의 콜백 함수가 실행됩니다.

단일 쓰레드의 장점:
싱글 쓰레드 모델은 쓰레드 간의 경쟁 조건 및 데드락과 같은 다중 쓰레드 프로그래밍에 따른 복잡성을 줄입니다. 이로 인해 Node.js 개발자는 비교적 간단한 코드로도 높은 동시성을 지원할 수 있습니다.

스케일링:
Node.js는 여러 프로세스 또는 클러스터링을 통해 멀티코어 CPU를 활용하여 더 많은 동시 요청을 처리할 수 있습니다. 이는 멀티플렉싱 및 비동기 처리를 통해 가능하며, 여러 Node.js 프로세스가 로드 밸런싱을 통해 요청을 분산 처리할 수 있습니다.

요약하면, Node.js는 비동기 이벤트 기반의 싱글 쓰레드 모델을 채택하여 높은 동시성을 제공하면서도 다중 쓰레드 프로그래밍의 복잡성을 피하기 위해 설계되었습니다. 이로 인해 I/O 집약적인 작업을 효율적으로 처리할 수 있고, 멀티코어 시스템에서도 스케일링이 가능합니다.

**Comment by javascriptbiggosoo** [comment]
웹 브라우저에서의 JavaScript 실행 환경과 Node.js의 실행 환경은 별개입니다. 그리고 자바스크립트를 지원하는 환경이 반드시 싱글 스레드로 동작해야 하는 것은 아닙니다. 실제로, 웹 브라우저와 Node.js 환경에서도 멀티스레딩을 지원하는 방법들이 있습니다. 웹 브라우저에서 JavaScript가 주로 싱글 스레드로 동작하는 것은, UI 렌더링과 상호작용이 주된 목적이기 때문에 동시성 문제를 피하기 위한 선택 중 하나입니다.

Node.js는 웹 브라우저 밖에서 동작하는 서버 사이드 환경에 최적화되어 설계된 JavaScript 런타임입니다. 그렇기 때문에, Node.js를 멀티 스레드로 설계할 수도 있었습니다. 실제로 Node.js에는 worker_threads 모듈이 있어 멀티 스레드 작업을 지원합니다.

그러나 Node.js의 초기 설계 철학은 비동기 I/O 작업에 중점을 둔 싱글 스레드, 이벤트 드리븐 아키텍처로 간단하고 효율적인 서버를 구축하는 것이었습니다. 이 아키텍처는 I/O 바운드 작업에 특히 잘 맞으며, 많은 네트워크 기반 애플리케이션에 적합하다고 판단되었습니다.


싱글 스레드 Node.js에서는 아래와 같이 간단하게 비동기 코드를 작성할 수 있습니다:

```javascript
const fs = require('fs');

// 비동기 파일 읽기
fs.readFile('example.txt', 'utf8', (err, data) => {
  if (err) throw err;
  console.log(data);
});

console.log('Reading file...');
```

이 코드는 'example.txt' 파일을 비동기적으로 읽고, 콘솔에 출력합니다. "Reading file..." 문자열은 파일 읽기가 완료되기 전에 출력됩니다.

만약 Node.js가 멀티스레드로 동작한다면, 스레드 관리와 동기화를 위한 추가적인 코드가 필요해집니다. 아래는 Java의 멀티스레드 코드 예시로, JavaScript에는 직접적으로 적용할 수 없지만, 멀티스레드 환경에서의 코드 복잡성을 대략적으로 보여줍니다:

```java
class ExampleThread extends Thread {
  public void run() {
    // 스레드에서 수행할 작업
    System.out.println("Thread is running.");
  }
}

public class Main {
  public static void main(String[] args) {
    ExampleThread thread1 = new ExampleThread();
    ExampleThread thread2 = new ExampleThread();
    
    // 스레드 시작
    thread1.start();
    thread2.start();
    
    // 메인 스레드에서 다른 작업 수행
    System.out.println("Main thread is running.");
  }
}
```

이 Java 예시에서는 여러 스레드를 생성, 시작하고 관리하는 등의 작업이 필요합니다. 동기화나 공유 자원 접근 문제 등을 해결하기 위한 추가적인 코드도 필요하게 됩니다. 이는 비교적 간단한 Node.js 코드에 비해 복잡성이 증가한 예시입니다.
